subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestClassProcessor.java

```diff
@@ -0,0 +1,206 @@
+import org.gradle.api.internal.tasks.testing.filter.TestSelectionMatcher;
+import org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor;
+import org.gradle.api.internal.tasks.testing.junit.TestClassExecutionListener;
+import org.gradle.internal.actor.ActorFactory;
+import org.gradle.internal.id.IdGenerator;
+import org.gradle.internal.time.Clock;
+import org.junit.platform.engine.DiscoverySelector;
+import org.junit.platform.engine.FilterResult;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestSource;
+import org.junit.platform.engine.discovery.DiscoverySelectors;
+import org.junit.platform.engine.support.descriptor.ClassSource;
+import org.junit.platform.engine.support.descriptor.MethodSource;
+import org.junit.platform.launcher.Launcher;
+import org.junit.platform.launcher.LauncherDiscoveryRequest;
+import org.junit.platform.launcher.PostDiscoveryFilter;
+import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder;
+import org.junit.platform.launcher.core.LauncherFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.gradle.api.internal.tasks.testing.junitplatform.VintageTestNameAdapter.*;
+import static org.junit.platform.launcher.EngineFilter.excludeEngines;
+import static org.junit.platform.launcher.EngineFilter.includeEngines;
+import static org.junit.platform.launcher.TagFilter.excludeTags;
+import static org.junit.platform.launcher.TagFilter.includeTags;
+
+public class JUnitPlatformTestClassProcessor extends AbstractJUnitTestClassProcessor<JUnitPlatformSpec> {
+    private TestResultProcessor resultProcessor;
+    private TestClassExecutionListener executionListener;
+    private CollectAllTestClassesExecutor testClassExecutor;
+
+    public JUnitPlatformTestClassProcessor(JUnitPlatformSpec spec, IdGenerator<?> idGenerator, ActorFactory actorFactory, Clock clock) {
+        super(spec, idGenerator, actorFactory, clock);
+    }
+
+    @Override
+    protected Action<String> createTestExecutor(TestResultProcessor threadSafeResultProcessor, TestClassExecutionListener threadSafeTestClassListener) {
+        resultProcessor = threadSafeResultProcessor;
+        executionListener = threadSafeTestClassListener;
+        testClassExecutor = new CollectAllTestClassesExecutor();
+        return testClassExecutor;
+    }
+
+    @Override
+    public void stop() {
+        testClassExecutor.processAllTestClasses();
+        super.stop();
+    }
+
+    private class CollectAllTestClassesExecutor implements Action<String> {
+        private final List<Class<?>> testClasses = new ArrayList<>();
+
+        @Override
+        public void execute(String testClassName) {
+            Class<?> testClass = loadClass(testClassName);
+            if (testClass != null && testClass.getEnclosingClass() == null) {
+                // Only process top level classes
```

Comment:
⭕️ Please don't use implementation comments like this. If something isn't clear, extract a method (e.g. `if(isTopLevelClass())`) or add some method-level or class-level documentation on what the general approach is.

Reply:
Now this part is totally removed. We don't need to filter top classes now because we process all classes in one pass instead of class by class and JUnit 5 can distinguish nested classes.

Reply:
We could still end up scheduling the top-level class on one worker and the nested class on another worker. That's actually very likely, because we use a round-robin algorithm and they would be alphabetically next to each other.

I think we'll have to add a workaround for this to our test class detection on the daemon side, so we don't schedule nested classes when the containing class is scheduled (or maybe not schedule them at all, not sure what the expected behavior should be). Can you open an issue for that investigation?

Reply:
Nice catch! I removed this just because I was not sure how to handle `ClassNotFoundException` here. 

Given this issue you mentioned, I'd bring this logic back. We don't need to load the classes on the daemon side just for checking if it's a nested class.

Reply:
Let's do that in a follow-up PR. This one is getting too large, it's hard to even open in the browser ;)